src/main/java/org/eolang/jeo/representation/CanonicalXmir.java

```diff
@@ -0,0 +1,120 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.eolang.jeo.representation;
+
+import com.jcabi.xml.XML;
+import com.yegor256.xsline.StClasspath;
+import com.yegor256.xsline.StEndless;
+import com.yegor256.xsline.TrClasspath;
+import com.yegor256.xsline.TrDefault;
+import com.yegor256.xsline.TrJoined;
+import com.yegor256.xsline.Xsline;
+import java.io.IOException;
+import org.cactoos.io.InputOf;
+import org.eolang.parser.EoSyntax;
+import org.eolang.parser.xmir.Xmir;
+
+/**
+ * Canonical XMIR.
+ * The purpose of this class is to restore the original XMIR format after PHI/UNPHI transformations.
+ * You can read more about it
+ * <a href="https://github.com/objectionary/eo/issues/3373#issuecomment-2361337359">here</a>.
+ * @since 0.6
+ */
+public final class CanonicalXmir {
+
+    /**
+     * Canonical XMIR after "phi/unphi" transformations.
+     */
+    private final XML canonical;
+
+    /**
+     * Constructor.
+     * @param canonical Significantly modified XMIR after "phi/unphi".
+     */
+    public CanonicalXmir(final XML canonical) {
+        this.canonical = canonical;
+    }
+
+    /**
+     * Convert canonical XMIR to plain XMIR.
+     * @return Plain XMIR.
+     */
+    public XML plain() {
+        try {
+            return CanonicalXmir.unroll(CanonicalXmir.parse(this.toEo()));
+        } catch (final IOException exception) {
+            throw new IllegalStateException(
+                "Can't parse the canonical XMIR",
+                exception
+            );
+        }
+    }
+
+    /**
+     * Unroll all the changes made by the "phi/unphi" transformations.
+     * @param parsed Parsed XMIR.
+     * @return Unrolled XMIR.
+     */
+    private static XML unroll(final XML parsed) {
+        return new Xsline(
+            new TrJoined<>(
+                new TrClasspath<>(
+                    "/org/eolang/parser/wrap-method-calls.xsl"
+                ).back(),
+                new TrDefault<>(
+                    new StEndless(
+                        new StClasspath(
+                            "/org/eolang/parser/roll-bases.xsl"
+                        )
+                    )
+                ),
+                new TrClasspath<>(
+                    "/org/eolang/parser/add-refs.xsl",
+                    "/org/eolang/parser/vars-float-down.xsl",
+                    "/org/eolang/parser/roll-data.xsl"
+                ).back()
+            )
+        ).pass(parsed);
+    }
+
+    /**
+     * Convert XMIR to EO.
+     * @return EO.
+     */
+    private String toEo() {
+        return new Xmir.Default(this.canonical).toEO();
+    }
+
+    /**
+     * Parse XMIR.
+     * @param eoprog Eo program.
+     * @return Parsed XMIR.
+     * @throws IOException If fails.
+     */
+    private static XML parse(final String eoprog) throws IOException {
+        return new EoSyntax("name", new InputOf(eoprog)).parsed();
```

Comment:
@@volodya-lombrozo I think this is the problem of https://github.com/objectionary/eo/issues/3373#issuecomment-2370887281

Reply:
@maxonfjvipon Please, read this issue https://github.com/objectionary/eo/issues/3380 it's a bug in `eo`

Reply:
@volodya-lombrozo I think the default constructor won't help you here because the `name` should taken from somewhere. Can't you set it manually?